mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,366 @@
+        ctx: &MmArc,
+        target_peer: PeerId,
+        is_a_reply: bool,
+        expires_in_seconds: i64,
+    ) -> Result<Self, String> {
+        let p2p_ctx = P2PContext::fetch_from_mm_arc(ctx);
+        let sender_peer = p2p_ctx.peer_id();
+        let keypair = p2p_ctx.keypair();
+        let sender_public_key = keypair.public().encode_protobuf();
+
+        let data = HealthcheckData {
+            sender_peer,
+            sender_public_key,
+            target_peer,
+            expires_at: Utc::now().timestamp() + expires_in_seconds,
+            is_a_reply,
+        };
+
+        let signature = try_s!(keypair.sign(&try_s!(data.encode())));
+
+        Ok(Self { signature, data })
+    }
+
+    pub(crate) fn is_received_message_valid(&self, my_peer_id: PeerId) -> bool {
+        let now = Utc::now().timestamp();
+        if now > self.data.expires_at {
+            log::debug!(
+                "Healthcheck message is expired. Current time in UTC: {now}, healthcheck `expires_at` in UTC: {}",
+                self.data.expires_at
+            );
+            return false;
+        }
+
+        if self.data.target_peer != my_peer_id {
+            log::debug!(
+                "`target_peer` doesn't match with our peer address. Our address: '{}', healthcheck `target_peer`: '{}'.",
+                my_peer_id,
+                self.data.target_peer
+            );
+            return false;
+        }
+
+        let Ok(public_key) = Libp2pPublic::try_decode_protobuf(&self.data.sender_public_key) else {
+            log::debug!("Couldn't decode public key from the healthcheck message.");
+
+            return false
+        };
+
+        if self.data.sender_peer != public_key.to_peer_id() {
+            log::debug!("`sender_peer` and `sender_public_key` doesn't belong each other.");
+
+            return false;
+        }
+
+        let Ok(encoded_message) = self.data.encode() else {
+            log::debug!("Couldn't encode healthcheck data.");
+            return false
+        };
+
+        let res = public_key.verify(&encoded_message, &self.signature);
+
+        if !res {
+            log::debug!("Healthcheck isn't signed correctly.");
+        }
+
+        res
+    }
+
+    #[inline]
+    pub(crate) fn encode(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> { encode_message(self) }
+
+    #[inline]
+    pub(crate) fn decode(bytes: &[u8]) -> Result<Self, rmp_serde::decode::Error> { decode_message(bytes) }
+
+    #[inline]
+    pub(crate) fn should_reply(&self) -> bool { !self.data.is_a_reply }
+
+    #[inline]
+    pub(crate) fn sender_peer(&self) -> PeerId { self.data.sender_peer }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[cfg_attr(any(test, target_arch = "wasm32"), derive(PartialEq))]
+struct HealthcheckData {
+    #[serde(deserialize_with = "deserialize_peer_id", serialize_with = "serialize_peer_id")]
+    sender_peer: PeerId,
+    #[serde(deserialize_with = "deserialize_bytes")]
+    sender_public_key: Vec<u8>,
```

Comment:
I guess we cannot recover the pubkey from the signature in this case?

Reply:
I don't think we can.